### ISSUE TITLE
Lazy init causes unexpected behavior?

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn_common#212]
I was writing this code:
```python
def test_param_name_deep():
  class _SubNet(nn.Module):
    def __init__(self, out_dim: nn.Dim):
      super().__init__()
      self.linear = nn.Linear(out_dim)

    def __call__(self, x) -> nn.Tensor:
      return self.linear(x)

  class _Net(nn.Module):
    def __init__(self):
      super().__init__()
      self.out_dim = nn.FeatureDim("linear-out", 13)
      self.sub = _SubNet(self.out_dim)
      self.sub2 = _SubNet(self.out_dim)
      self.sub2.linear.weight = self.sub.linear.weight
      self.sub2.linear.bias = self.sub.linear.bias

    def __call__(self, x: nn.Tensor) -> nn.Tensor:
      return self.sub(x) + self.sub2(x)

  net = _Net()
  config, net_dict = dummy_config_net_dict(net)
  pprint(config)
  dummy_run_net(config, net=net)
```
Note the intended param sharing of `sub` an `sub2`.

This code does not produce any error, but it does not share the params! This was somehow unexpected.

The reason is the lazy init of the params when the input dim is not specified, for `nn.Linear` here. (Related: #75)
At this point, `weight` and `bias` in `_Net.__init__` are in fact `None` and this just does nothing.
The params are only set after the first `__call__`.

This however now makes the param sharing somewhat annoying. My code looks now like this:
```python
  class _Net(nn.Module):
    def __init__(self):
      super().__init__()
      self.out_dim = nn.FeatureDim("linear-out", 13)
      self.sub = _SubNet(self.out_dim)
      self.sub2 = _SubNet(self.out_dim)

    def __call__(self, x: nn.Tensor) -> nn.Tensor:
      y = self.sub(x)
      # Can only assign params now due to lazy init...
      self.sub2.linear.initialized = True
      self.sub2.linear.in_dim = x.feature_dim
      self.sub2.linear.weight = self.sub.linear.weight
      self.sub2.linear.bias = self.sub.linear.bias
      return y + self.sub2(x)
```
So I had to move the param sharing logic over to the `__call__`, right after the first `sub(x)` call.
I don't really like this.

So, the question: Should we remove the lazy init logic? Or anything else we should do?
(@JackTemaki @Atticus1806)
(Related: #149)

**Comment by albertz** [comment]
Another question: How does PyTorch handle this situation? They also have lazy modules, although they keep them separate, e.g. [`LazyConv2d`](https://pytorch.org/docs/stable/generated/torch.nn.LazyConv2d.html#torch.nn.LazyConv2d). See [here](https://pytorch.org/docs/stable/generated/torch.nn.modules.lazy.LazyModuleMixin.html#torch.nn.modules.lazy.LazyModuleMixin).

**Comment by albertz** [comment]
@JackTemaki @Atticus1806 opinions here?

**Comment by Atticus1806** [comment]
This is definitely unexpected. But I am not sure how to handle this. Maybe via some warning when this happens?

**Comment by albertz** [comment]
We already have the warning, but the warning is not triggered in this example code. And there will always be some potential code where no warning would be triggered, so it's always potentially dangerous.

Despite that, it's also unintuitive and complicated, once you want to do any handling on parameters, like weight norm, weight dropout, L2, sharing, etc, because you always would first need to feed some input through the model.

Despite that, the lazy init logic in `Linear` etc also make the `Linear` code way more complicated. Most of the current `Linear` code is just about that.

After some internal discussion, we (@JackTemaki @Atticus1806 me) agree that it's better to completely remove the lazy init logic. A user could in principle still do sth like that in custom modules, but none of the default modules will have that.

**Comment by albertz** [comment]
Note that the order of positional args should probably also change. Currently we have:
```python
rc.nn.Linear(out_dim: nn.Dim, *, in_dim: Optional[nn.Dim] = None, ...)
```
Currently `in_dim` is named arg only. But when we change this, we should also allow it to be positional, and then overtake the PyTorch order, to make it consistent to PyTorch. Specifically:
```python
torch.nn.Linear(in_features, out_features, ...)
```
I.e. for us:
```python
rc.nn.Linear(in_dim: nn.Dim, out_dim: nn.Dim, *, ...)
```
It's not really possible that any current code would unexpectedly break. Either it breaks with some exception, or when you put named args it will already be correct.

**Comment by albertz** [comment]
TODO:

- [ ] check `check_in_feature_dim_lazy_init` usages. Those are all for lazy init. In the end, remove that function.
- [ ] `nn.Module.initialized` attrib is for lazy init. remove that.

Specifically, the relevant modules:

- `Linear`
- `_Conv`
- `_TransposedConv`
- `BatchNorm`
- `LayerNorm`

**Comment by albertz** [comment]
While removing the lazy init logic, I now define the param random init directly in the module `__init__`. However, now I run into #109. So this needs to be fixed as well.
